core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -135,52 +141,53 @@ public void start() {
+            nonTxWriters.forEach(w -> startWriter(w, undelegated));
+            txWriters.forEach(w -> startWriter(w, undelegated));
+            loaders.forEach(l -> startLoader(l, undelegated));
 
-            Set<Lifecycle> undelegated = new HashSet<>();//black magic to make sure the store start only gets invoked once
+            // Ensure that after writers and loaders have started, they are classified as available by their isAvailable impl
+            pollStoreAvailability();
 
-            Consumer<CacheWriter> startWriter = writer -> {
-               writer.start();
-               if (writer instanceof DelegatingCacheWriter) {
-                  CacheWriter actual = undelegate(writer);
-                  actual.start();
-                  undelegated.add(actual);
-               } else {
-                  undelegated.add(writer);
-               }
-
-               if (configMap.get(writer).purgeOnStartup()) {
-                  if (!(writer instanceof AdvancedCacheWriter))
-                     throw new PersistenceException("'purgeOnStartup' can only be set on stores implementing " +
-                                                          "" + AdvancedCacheWriter.class.getName());
-                  ((AdvancedCacheWriter) writer).clear();
-               }
-            };
-            nonTxWriters.forEach(startWriter);
-            txWriters.forEach(startWriter);
-
-            for (CacheLoader l : loaders) {
-               if (!undelegated.contains(l))
-                  l.start();
-               if (l instanceof DelegatingCacheLoader) {
-                  CacheLoader actual = undelegate(l);
-                  if (!undelegated.contains(actual)) {
-                     actual.start();
-                  }
-               }
-            }
+            // Now schedule the availability check
+            long interval = configuration.persistence().availabilityInterval();
+            persistenceExecutor.scheduleAtFixedRate(this::pollStoreAvailability, interval, interval, TimeUnit.MILLISECONDS);
          } finally {
             if (xaTx != null) {
                transactionManager.resume(xaTx);
             }
-            storesMutex.readLock().unlock();
+            storesMutex.writeLock().unlock();
          }
       } catch (Exception e) {
          throw new CacheException("Unable to start cache loaders", e);
       }
    }
 
+   private void pollStoreAvailability() {
+      storesMutex.writeLock().lock();
+      try {
+         for (StoreStatus status : storeStatuses.values()) {
+            if (!status.availabilityChanged())
+               continue;
+
+            cacheNotifier.notifyPersistenceAvailabilityChanged(status.availability);
+            if (!status.availability) {
+               unavailableException = new StoreUnavailableException(String.format("Store %s is unavailable", status.store));
+               managerAvailable = false;
+               return;
```

Comment:
Should we check all of the stores availability, even when one fails?

Reply:
Probably ... I was thinking that there is no need to check all stores availability if one is unavailable, as that is enough for the PersistenceManager to become unavailable. However, I guess it makes more sense to keep track of all store statuses so that the PM can become available sooner in the event of the unavailable store being removed from the PM.